third_party/move/move-vm/runtime/src/storage/layout_cache.rs

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! Data structures for caching layouts of VM values ina long-living, concurrent cache. The design
+//! goal is to make sure that cache hit is semantically equivalent to layout construction. That is,
+//! if layout construction costs 5 units of gas, so does the cache hit. Failed layout construction
+//! never results in insertion into the cache.
+//!
+//! Additionally, caches store only roots of the layout tree. There is NO caching for sub-layouts
+//! because:
+//!   1. This is more error-prone because enforcing of semantic equivalence when reading sub-layout
+//!      is more difficult: e.g., one needs to ensure the depth and size of layouts are correct.
+//!   2. Arguably, we need layouts for root-like values only (e.g., those with `key` ability).
```

Comment:
root == resource, it appears. 

I think you may be loosing optimization potential if only caching resources as a whole. For example, one can imagine complex layouts for `OrderedMap` appearing with the same instantations many times.

However, I recognize the problem of not having Arc in the layout type itself, so this would be overkill for this PR.

Otherwise, shouldn't be too hard, because mechanism for arbitrary cached struct layout gas cost should be similar as for 'roots'.